fix: add cluster api with ipv6 appending '[' in secret name (#8204)

zillani opened the pull request:
**URIToSecretName** method hashes a URI address and returns a secret name
- if ipv6 URI is provided, then the method is appending '[' in secret name 
- this is causing an error like Secret "cluster-[fcff-1931808423" is invalid 
- fixed by trimming '[' from the string

Signed-off-by: zillani <shaikzillani@gmail.com>

**Edit**
- Parsed & Added IPV6 hostip to secret name, this will avoid ipv6 hash collisions. 
- Added test cases

This is working for, 
- hostname & hostname:<port>
- IPv4address & IPv4address:<port>
- IPv6address & IPv6address:<port>

This PR address IPv6 scenario without impacting existing functionality.

zillani commented:
fixes #11319

jessesuen reviewed (changes requested):
Could you please add a unit test for this? Also, is it possible that replacing all occurrences of `[` with `` cause potential collision in secret name?

zillani commented:
yea, I am working on it, any suggestions on making the secret unique?

zillani commented:
I have just added the test case & also added ipv6 hostname to the string, this will avoid any hash collisions, 

**Format**
A secret name will be created in this format, 
cluster-<IP or Hostname>-HASH

For hashing, argo is already using the **FNV1a** algorithm with 32bit hash which has lesser collisions compared to other algorithms. However, since we are appending the hostname to the secret, these collisions are near impossible. 

**Ipv4**
Input URI -> http://127.0.0.1:8080
Secret name -> cluster-127.0.0.1-2348797202

**Ipv6**
Input URI -> **http://[fcff:69:90:0:5054:ff:fea7:810]:8080**
Secret name -> **cluster-fcff699005054fffea78108080-4259157202**

Here, removing **:** instead of replacing is better, double **:** may occur in ipv6 & extra **:** if port is present

zillani commented:
Hi @jessesuen any update?

zillani commented:
Hi @jessesuen any update on this?

zillani commented:
@jessesuen can you please review

tossmilestone commented on util/db/secrets.go:
```diff
@@ -158,5 +158,18 @@ func URIToSecretName(uriType, uri string) (string, error) {
+	parsedHost := strings.Split(parsedURI.Host, ":")
+
+	//If IPv6 address with port
+	if len(parsedHost) >= 8 {
```
Why use `len(parsedHost) >= 8` to determine if it's ipv6 address? There is already a `ParseIP` function in  `net` package to do the work.

tossmilestone commented:
Hi, @jessesuen Can you help review this PR? We need ArgoCD to support IPv6 clusters. Thank you！

zillani commented:
> Hi, @jessesuen Can you help review this PR? We need ArgoCD to support IPv6 clusters. Thank you！

hey @tossmilestone  will work on this today

tossmilestone commented:
> > Hi, @jessesuen Can you help review this PR? We need ArgoCD to support IPv6 clusters. Thank you！
> 
> 
> 
> hey @tossmilestone  will work on this today

@zillani hi，is the pr still woking in process?
